lessons/specifics/plug.md

```diff
@@ -51,17 +58,117 @@ defmodule HelloWorldPlug do
+Since we're starting a Plug application from scratch, we need to define the application module.
+Update `lib/example.ex` to start and supervise Cowboy:
 
-_Note_: Plugs are applied to all requests which is why we will handle filtering requests and applying our logic to only a subset of them.  To ignore a request we simply pass the connection through.
+```elixir
+defmodule Example do
+  use Application
+  require Logger
 
-We'll start by looking at our finished Plug and then discuss how it works.  We'll create it at `lib/plug/verify_request.ex`:
+  def start(_type, _args) do
+    children = [
+      Plug.Adapters.Cowboy.child_spec(:http, Example.HelloWorldPlug, [], port: 8080)
+    ]
+
+    Logger.info "Started application"
+
+    Supervisor.start_link(children, strategy: :one_for_one)
+  end
+end
+```
+
+This supervises Cowboy, and in turn, supervises our `HelloWorldPlug`.
+
+Now, the `application` portion of `mix.exs` needs two things:
+1) A list of dependency applications (`cowboy`, `logger`, and `plug`) that need to start up, and
+2) Configuration for our own application, which should also start up automatically.
+Let's update the `application` portion of `mix.exs` to do that:
+
+```elixir
+def application do
+  [
+    applications: [:cowboy, :logger, :plug],
+    mod: {Example, []}
+  ]
+end
+```
+
+We're ready to try out this minimalistic, plug-based web server.
+On the command line, run:
+
+```shell
+$ mix run --no-halt
+```
+
+Once everything is finished compiling, and `[info]  Started app` appears, open a web
+browser to `localhost:8080`. It should display:
+
+```
+Hello World!
+```
+
+## Plug.Router
+
+For most applications, like a web site or REST API, you'll want a router to route request for different paths and HTTP verbs to different handlers.
+`Plug` provides a router to do that. As we are about to see, we don't need a framework like Sinatra in Elixir since we get that for free with Plug.
+
+To start let's create a file at `lib/example/router.ex` and copy the following into it:
+
+```elixir
+defmodule Example.Router do
+  use Plug.Router
+
+  plug :match
+  plug :dispatch
+
+  get "/", do: send_resp(conn, 200, "Welcome")
+  match _, do: send_resp(conn, 404, "Oops!")
+end
+```
+
+This is a bare minimum Router but the code should be pretty self-explanatory.
+We've included some macros through `use Plug.Router` and then set up two of the built-in Plugs: `:match` and `:dispatch`.
+There are two defined routes, one for handling GET requests to the root and the second for matching all other requests so we can return a 404 message.
+
+Back in `mix.exs`, we need to tell Elixir about our router.
+Swap out the `Example.HelloWorldPlug` plug with the new router:
+
+```elixir
+def application do
+  [
+    applications: [:cowboy, :logger, :plug],
+    mod: {Example.Router, []}
```

Comment:
I find that changing the Application Module Callback to `Example.Router` results in a Mix error because the compiler is expecting the `start/2` callback to start the supervision tree, but the callback remains in `lib/example.ex`. Instead, I changed the plug given to `Plug.Adapters.Cowboy.child_spec/4` in `lib/example.ex` from `Example.HelloWorldPlug` to `Example.Router`. This is also consistent with the rest of the lesson (see the Making the HTTP Port Configurable sub-lesson)